### PR TITLE
Gulpless

### PR DESCRIPTION
**Pull request by pspeter3**
This is to remove our usage of gulp. One downside so far has been the fact that `prepublish` is run on `npm install` so a lot of the before deploy work needs to happen in Travis.

### CODE REVIEW

File: src/perishable.ts [hunk 1]
```diff
@@ -54,8 +54,10 @@ class PerishableNode<T> implements Handle<T> {
         if (this._hasNext()) {
             this._next._setPrev(this._prev);
         }
-        // Perishable nodes should only be released once but in case we strip assertions we want to make sure that we
-        // don't accidentally add _next into _prev's list if we are removed twice.
+        /**
+         * Perishable nodes should only be released once but in case we strip assertions we want to make sure that we
```
Comment (verticalpalette) [inline src/perishable.ts:58]: I actually prefer single line comments. They're easier for humans (or at least me) to type, that's how the editors I use create comments, and they can be nested, unlike block comments.

File: tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,47 @@
+{
+    "version": "1.5.0-beta",
```
Comment (verticalpalette) [inline tsconfig.json:2]: Any potential issues with different versions of tsc?

File: README.md [hunk 3]
```diff
@@ -17,10 +19,6 @@ Install with npm
 npm install tsutil --save
 ```
 
-## Documentation
```
Comment (verticalpalette) [inline README.md:20]: We should update the installation instructions too. It's missing the `npm run typings` step.

File: package.json [hunk 4]
```diff
@@ -8,9 +8,15 @@
     "*.d.ts"
   ],
   "scripts": {
-    "bundle": "gulp bundle",
-    "doc": "gulp doc",
-    "test": "gulp test"
+    "bundle": "npm run clean && npm run compile && mv _build/src/** . && node _build/tools/bundle.js",
```
Comment (verticalpalette) [inline package.json:11]: Doesn't seem much better than a Makefile, tbh. Also, I expect this to be completely unapproachable to someone new -- e.g. what command should I run to do X? It's a shame JSON doesn't support comments.

File: package.json [hunk 5]
```diff
@@ -8,9 +8,15 @@
...
+    "clean": "rm -rf _build coverage doc",
+    "compile": "tsc",
+    "coverage": "cat ./coverage/lcov.info | coveralls",
+    "lint": "find src test -name \"*.ts\" | sed 's/^/--file=/g' | xargs tslint",
+    "pretest": "npm run compile -- --sourceMap & npm run lint",
+    "test": "istanbul cover _mocha -- --reporter $npm_package_config_reporter --slow 10 --ui bdd --recursive _build/test",
```
Comment (verticalpalette) [inline package.json:17]: Does the `$npm_package_config_reporter` variable come in the environment for free? Also, want to put quotes or braces around it? I can see it doesn't have spaces in it, but if it did (or if someone copied/pasted this command)...

File: package.json [hunk 6]
```diff
@@ -40,24 +46,22 @@
...
-    "gulp-istanbul": "^0.5.0",
-    "gulp-mocha": "^2.0.0",
-    "gulp-replace": "^0.5.3",
-    "gulp-tslint": "^1.4.3",
-    "gulp-typedoc": "^1.0.6",
-    "gulp-typescript": "^2.4.2",
-    "sinon": "^1.12.2",
-    "typescript-formatter": "^0.1.4",
-    "typescript-require": "pspeter3/typescript-require"
+    "istanbul": "^0.3.13",
```
Comment (verticalpalette) [inline package.json:52]: It's nice that we can get rid of so many dependencies.

File: tsconfig.json [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    },
+    "compileOnSave": false,
+    "filesGlob": [
+        "./**/*.ts",
+        "!./node_modules/**/*.ts",
+        "!./_build/**/*.ts",
+        "!./*.ts"
+    ],
+    "files": [
+        "./src/arrayFind.ts",
```
Comment (verticalpalette) [inline tsconfig.json:18]: Seems like files and filesGlob are redundant. Why have both?

**Comment by verticalpalette** [comment]
lgtm, with the caveat that we revert if this gets hairy.

File: src/perishable.ts (hunk 1 above)
Comment (pspeter3) [inline src/perishable.ts:58]: This is a result of our lint rules, I'm happy to change them but I figured that was a separate commit.

File: tsconfig.json (hunk 2 above)
Comment (pspeter3) [inline tsconfig.json:2]: Not that I can see. We could use `1.4.2` if we really wanted too.

File: README.md (hunk 3 above)
Comment (pspeter3) [inline README.md:20]: I think this should be in the development section.

File: package.json (hunk 4 above)
Comment (pspeter3) [inline package.json:11]: Yeah, I wish there were JSON comments as well.

File: package.json (hunk 5 above)
Comment (pspeter3) [inline package.json:17]: It comes from the `config` section. I tried avoiding using that feature but it was difficult to have different reporters for local and travis.

File: package.json (hunk 6 above)
Comment (pspeter3) [inline package.json:52]: Yeah, that was the most exciting part. It also seems like it runs faster too now.

File: tsconfig.json (hunk 7 above)
Comment (pspeter3) [inline tsconfig.json:18]: `filesGlob` is for `atom-typescript`. My other goal is to write a tool which will look for that key and then regenerate the `files` array. It's extremely annoying to maintain the files array by hand. TypeScript supports not using it but lots of tools fill it in automatically.

**Comment by pspeter3** [comment]
Yeah, I'd be down to revert if we think this was a bad idea. Some ideas I've seen online include using specific npm command line tools like `rimraf` instead of `rm -rf` so that way you don't rely on OS. I'm planning on following up with more PRs that simplify the commands.

**Pull request merged**